src/librustc_typeck/check/cast.rs

```diff
@@ -345,7 +346,7 @@ impl<'tcx> CastCheck<'tcx> {
     {
         // array-ptr-cast.
 
-        if m_expr.mutbl == hir::MutImmutable && m_cast.mutbl == hir::MutImmutable {
+	if !(m_expr.mutbl == hir::MutImmutable && m_cast.mutbl == hir::MutMutable) {
```

Comment:
I don't think this is correct - this would only allow casts from `&[T; n]` to `*mut T`, whereas you want casts from immutable to const and mut to mut.